src/Compiler/Facilities/AsyncMemoize.fs

```diff
@@ -1,19 +1,85 @@
 namespace Internal.Utilities.Collections
 
 open System
-open System.Collections.Generic
 open System.Diagnostics
 open System.IO
 open System.Threading
 open System.Threading.Tasks
 
-open FSharp.Compiler
-open FSharp.Compiler.BuildGraph
-open FSharp.Compiler.Diagnostics
 open FSharp.Compiler.DiagnosticsLogger
 open Internal.Utilities.Library
 open System.Runtime.CompilerServices
 
+type AsyncLazyState<'t> =
+    | Initial of Async<'t>
+    | Created of Task<'t> * CancellationTokenSource * int
+
+/// Represents a computation that will execute only once but can be requested by multiple clients.
+/// It keeps track of the number of requests. When all clients cancel their requests, the underlying computation will also cancel and can be restarted.
+/// If cancelUnawaited is set to false, the computation will run to completion even when all requests are canceled.
+type AsyncLazy<'t>(computation: Async<'t>, ?cancelUnawaited: bool) =
+
+    let cancelUnawaited = defaultArg cancelUnawaited true
+
+    let stateUpdateSync = obj()
+    let mutable state = Initial computation
+
+    let cancelIfUnawaited () =
+        match state with
+        | Created(work, cts, 0) when not work.IsCompleted ->
+            cts.Cancel()
+            state <- Initial computation
+        | _ -> ()
+
+    let afterRequest () =
+        match state with
+        | Created(work, cts, count) ->
+            state <- Created(work, cts, count - 1)
+            if cancelUnawaited then cancelIfUnawaited () 
+        | _ -> ()
```

Comment:
Can this happen? Should we add a `failwith` to see if we get some unexpected states?